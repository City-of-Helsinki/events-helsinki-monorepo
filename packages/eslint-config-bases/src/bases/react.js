/**
 * Opinionated config base for projects using react.
 * @see https://github.com/City-of-Helsinki/events-helsinki-monorepo/tree/main/packages/eslint-config-bases
 */

import react from 'eslint-plugin-react';
import reactHooks from 'eslint-plugin-react-hooks';

const reactPatterns = {
  files: ['**/*.{js,jsx,ts,tsx}'],
};

export default [
  {
    files: reactPatterns.files,
    plugins: {
      react,
    },
    settings: {
      react: {
        version: 'detect',
      },
    },
    rules: {
      ...react.configs.recommended.rules,
      // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-unknown-property.md
      'react/no-unknown-property': ['error', { ignore: ['css'] }],
      // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unescaped-entities.md
      'react/no-unescaped-entities': ['error', { forbid: ['>'] }],
      'react/prop-types': 'off',
      'react/react-in-jsx-scope': 'off',
      // Fine-tune naming convention react typescript jsx (function components)
      // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/naming-convention.md
      '@typescript-eslint/naming-convention': [
        'warn',
        {
          selector: 'variable',
          format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
          leadingUnderscore: 'allow',
        },
        {
          selector: ['function'],
          format: ['camelCase', 'PascalCase'],
        },
        {
          selector: 'parameter',
          format: ['camelCase', 'PascalCase'],
          leadingUnderscore: 'allow',
        },
      ],
    },
  },
  {
    files: reactPatterns.files,
    plugins: {
      'react-hooks': reactHooks,
    },
    rules: {
      ...reactHooks.configs.recommended.rules,
      //   'react-hooks/rules-of-hooks': 'error',
      //   'react-hooks/exhaustive-deps': 'warn',
    },
  },
];
