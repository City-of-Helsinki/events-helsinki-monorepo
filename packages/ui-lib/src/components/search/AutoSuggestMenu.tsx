import type { AutosuggestMenuOption } from '@events-helsinki/core-lib/types';
import classNames from 'classnames';
import { IconCross } from 'hds-react';
import { useTranslation } from 'next-i18next';
import type { FunctionComponent } from 'react';
import React from 'react';

import styles from './autosuggestMenu.module.scss';

interface Props {
  focusedOption: number;
  isOpen: boolean;
  onClose: () => void;
  onOptionClick: (item: AutosuggestMenuOption) => void;
  options: AutosuggestMenuOption[];
}

const AutoSuggestMenu: FunctionComponent<Props> = ({
  focusedOption,
  isOpen,
  onClose,
  onOptionClick,
  options,
}) => {
  const { t } = useTranslation();

  if (!isOpen) return null;

  return (
    <div className={styles.autosuggestMenu}>
      <div className={styles.title}>
        {t('common:autosuggest.menu.title')}
        <button
          aria-label={t('common:autosuggest.menu.ariaButtonClose')}
          className={styles.closeButton}
          onClick={onClose}
        >
          <IconCross aria-hidden />
        </button>
      </div>
      <ul className={styles.autosuggestOptions} role="listbox">
        {options.map((option, index) => {
          const handleClick = () => {
            onOptionClick(option);
          };
          const handleKeyDown = (event: React.KeyboardEvent<HTMLLIElement>) => {
            // check keys if you want
            if (event.keyCode == 13) {
              handleClick();
            }
          };

          return (
            <li
              key={index}
              className={classNames(
                styles.autosuggestOption,
                styles[`autosuggestOption--${option.type}`],
                {
                  [styles['autosuggestOption--isFocused']]:
                    focusedOption === index,
                }
              )}
              role="option"
              tabIndex={-1}
              onClick={handleClick}
              onKeyDown={handleKeyDown}
              aria-selected={focusedOption === index}
            >
              <div className={styles.text}>{option.text}</div>
            </li>
          );
        })}
      </ul>

      <div className={styles.info}>{t('common:autosuggest.menu.info')}</div>
    </div>
  );
};

export default AutoSuggestMenu;
