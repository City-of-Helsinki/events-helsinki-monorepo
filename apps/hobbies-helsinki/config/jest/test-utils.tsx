import type { ApolloCache, InMemoryCache } from '@apollo/client/cache';
import type { MockedResponse } from '@apollo/client/testing';
import type { RenderResult } from '@testing-library/react';
import { act, fireEvent, render } from '@testing-library/react';
import type { NextRouter } from 'next/router';
import Router from 'next/router';
import type { ReactElement } from 'react';
import React from 'react';
import wait from 'waait';

import { hobbiesApolloClient } from '../../src/domain/clients/hobbiesApolloClient';
import TestProviders from './TestProviders';

type CustomRender = {
  (
    ui: React.ReactElement,
    options?: {
      mocks?: MockedResponse[];
      cache?: ApolloCache<Record<string, unknown>> | InMemoryCache;
      routes?: string | string[];
    }
  ): CustomRenderResult;
};

export type CustomRenderResult = RenderResult & { router: NextRouter };

export const arrowUpKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 38, key: 'ArrowUp' });

export const arrowDownKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 40, key: 'ArrowDown' });

export const enterKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 13, key: 'Enter' });

export const escKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 27, key: 'Escape' });

export const tabKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 9, key: 'Tab' });

const customRender: CustomRender = (
  ui: ReactElement,
  { mocks = [], cache = hobbiesApolloClient.cache, routes = [] } = {}
) => {
  if (routes) {
    if (!Array.isArray(routes)) {
      routes = [routes];
    }
    routes.forEach((route) => Router.push(route));
  }

  // set locales so that the routing tests for default locale worked
  Router.locales = ['fi', 'en', 'sv'];

  const renderResult = render(ui, {
    wrapper: ({ children }) => (
      <TestProviders mocks={mocks} router={Router} cache={cache}>
        {children}
      </TestProviders>
    ),
  });
  return {
    ...renderResult,
    router: Router,
  };
};

// eslint-disable-next-line testing-library/no-unnecessary-act
const actWait = (amount?: number): Promise<void> => act(() => wait(amount));

// eslint-disable-next-line import/export
export { actWait, customRender as render };

// re-export everything
// eslint-disable-next-line import/export
export * from '@testing-library/react';
export { render as defaultRender } from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
