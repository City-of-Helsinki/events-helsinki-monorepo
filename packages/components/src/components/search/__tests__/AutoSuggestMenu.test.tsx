import { AUTOSUGGEST_TYPES } from 'events-helsinki-core';
import * as React from 'react';

import { render } from '@/test-utils';
import AutoSuggestMenu from '../AutoSuggestMenu';
it('AutosuggestMenu matches snapshot', () => {
  const { container } = render(
    <AutoSuggestMenu
      focusedOption={0}
      options={[{ text: 'foo', type: AUTOSUGGEST_TYPES.TEXT, value: 'foo' }]}
      isOpen={true}
      onClose={jest.fn()}
      onOptionClick={jest.fn()}
    />
  );

  expect(container.firstChild).toMatchSnapshot();
});
