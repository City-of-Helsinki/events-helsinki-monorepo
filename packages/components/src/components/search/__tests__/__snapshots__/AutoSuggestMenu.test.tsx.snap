// Vitest Snapshot v1

exports[`AutosuggestMenu matches snapshot 1`] = `
<div
  class="_autosuggestMenu_bd6532"
>
  <div
    class="_title_bd6532"
  >
    autosuggest.menu.title
    <button
      aria-label="autosuggest.menu.ariaButtonClose"
      class="_closeButton_bd6532"
    >
      <svg
        aria-hidden="true"
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        role="img"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <rect
            height="24"
            width="24"
          />
          <polygon
            fill="currentColor"
            points="18 7.5 13.5 12 18 16.5 16.5 18 12 13.5 7.5 18 6 16.5 10.5 12 6 7.5 7.5 6 12 10.5 16.5 6"
          />
        </g>
      </svg>
    </button>
  </div>
  <ul
    class="_autosuggestOptions_bd6532"
    role="listbox"
  >
    <li
      aria-selected="true"
      class="_autosuggestOption_bd6532 _autosuggestOption--text_bd6532 _autosuggestOption--isFocused_bd6532"
      role="option"
      tabindex="-1"
    >
      <div
        class="_text_bd6532"
      >
        foo
      </div>
    </li>
  </ul>
  <div
    class="_info_bd6532"
  >
    autosuggest.menu.info
  </div>
</div>
`;
